lib/util/github/graphql/datasource-helper.ts

```diff
@@ -0,0 +1,256 @@
+  GithubGraphqlDatasourceAdapter,
+  GithubGraphqlPayload,
+  GithubGraphqlRepoParams,
+  GithubGraphqlRepoResponse,
+  GithubPackageConfig,
+  RawQueryResponse,
+} from './types';
+
+function isSomethingWentWrongMessage(msg: string): boolean {
+  return msg.startsWith('Something went wrong while executing your query.');
+}
+
+/**
+ * See: #16343
+ */
+function isErrorFromIssue16343(err: Error): boolean {
+  if (err instanceof AggregateError) {
+    const errors = [...(err as AggregateError<Error>)];
+    const messages = errors.map(({ message }) => message);
+    return messages.some(isSomethingWentWrongMessage);
+  }
+
+  return isSomethingWentWrongMessage(err.message);
+}
+
+export class GithubGraphqlDatasourceHelper<
+  GraphqlItem,
+  ResultItem extends GithubDatasourceItem
+> {
+  static prepareQuery(payloadQuery: string): string {
+    return `
+      query($owner: String!, $name: String!, $cursor: String, $count: Int!) {
+        repository(owner: $owner, name: $name) {
+          isRepoPrivate: isPrivate
+          payload: ${payloadQuery}
+        }
+      }
+    `;
+  }
+
+  static async query<T, U extends GithubDatasourceItem>(
+    config: GithubPackageConfig,
+    http: GithubHttp,
+    adapter: GithubGraphqlDatasourceAdapter<T, U>
+  ): Promise<U[]> {
+    const instance = new GithubGraphqlDatasourceHelper<T, U>(
+      config,
+      http,
+      adapter
+    );
+    const items = await instance.getItems();
+    return items;
+  }
+
+  private readonly baseUrl: string;
+  private readonly repoOwner: string;
+  private readonly repoName: string;
+
+  private itemsPerQuery: 100 | 50 | 25 = 100;
+
+  private queryCount = 0;
+
+  private cursor: string | null = null;
+
+  constructor(
+    packageConfig: GithubPackageConfig,
+    private http: GithubHttp,
+    private datasourceAdapter: GithubGraphqlDatasourceAdapter<
+      GraphqlItem,
+      ResultItem
+    >
+  ) {
+    const { packageName, registryUrl } = packageConfig;
+    [this.repoOwner, this.repoName] = packageName.split('/');
+    this.baseUrl = getApiBaseUrl(registryUrl).replace(/\/v3\/$/, '/'); // Replace for GHE
+  }
+
+  private getFingerprint(): string {
+    return [
+      this.baseUrl,
+      this.repoOwner,
+      this.repoName,
+      this.datasourceAdapter.key,
+    ].join(':');
+  }
+
+  private getRawQueryOptions(): GithubHttpOptions {
+    const baseUrl = this.baseUrl;
+    const repository = `${this.repoOwner}/${this.repoName}`;
+    const query = this.datasourceAdapter.query;
+    const variables: GithubGraphqlRepoParams = {
+      owner: this.repoOwner,
+      name: this.repoName,
+      count: this.itemsPerQuery,
+      cursor: this.cursor,
+    };
+
+    return {
+      baseUrl,
+      repository,
+      body: { query, variables },
+    };
+  }
+
+  private async doRawQuery(): Promise<
+    RawQueryResponse<GithubGraphqlPayload<GraphqlItem>>
+  > {
+    const requestOptions = this.getRawQueryOptions();
+
+    type HttpBody = GithubGraphqlResponse<
```

Comment:
double nested type? maybe on wrong type?

Reply:
Outer type is from upper-level `http` types, inner one is specific to GraphQL datasource